### PR TITLE
Python: fix Google AI/Vertex AI crash on anyOf schema (#12442)

### LINKED ISSUE #12442: Bug: ValueError: Unknown field for Schema: anyOf
Describe the bug
When using ChatCompletionAgent instances as plugins with GoogleAIChatCompletion service, the application crashes with ValueError: Unknown field for Schema: anyOf during schema conversion from Semantic Kernel to Google AI's protobuf format.
To Reproduce

Create multiple ChatCompletionAgent instances using GoogleAIChatCompletion service
Add these agents as plugins to another ChatCompletionAgent using the plugins parameter
Call get_response() method on the parent agent with any message
See error: ValueError: Unknown field for Schema: anyOf

Minimal reproduction code:
pythonimport asyncio
from semantic_kernel.agents import ChatCompletionAgent
from semantic_kernel.connectors.ai.google.google_ai import GoogleAIChatCompletion

# Initialize Google AI chat completion service
chat_completion = GoogleAIChatCompletion(
    gemini_model_id="gemini-2.0-flash",
    api_key="YOUR_API_KEY"
)

# Create child agents
billing_agent = ChatCompletionAgent(
    service=chat_completion, 
    name="BillingAgent", 
    instructions="You handle billing issues."
)

refund_agent = ChatCompletionAgent(
    service=chat_completion,
    name="RefundAgent",
    instructions="Assist users with refund inquiries.",
)

# Create parent agent with child agents as plugins
triage_agent = ChatCompletionAgent(
    service=chat_completion,
    name="TriageAgent",
    instructions="Evaluate user requests and forward them to other agents.",
    plugins=[billing_agent, refund_agent],  # This causes the error
)

async def main():
    # This call triggers the error
    response = await triage_agent.get_response(
        messages="test message",
        thread=None,
    )

asyncio.run(main())
Expected behavior
The triage agent should successfully process the user input and coordinate with the plugin agents to provide a response without any schema-related errors.
Screenshots
Full error traceback:
Traceback (most recent call last):
  File "C:\gen-ai\agents_poc\semantic_kernel\venv\Lib\site-packages\proto\marshal\rules\message.py", line 36, in to_proto
    return self._descriptor(**value)
           ~~~~~~~~~~~~~~~~^^^^^^^^^
TypeError: Parameter to CopyFrom() must be instance of same class: expected <class 'Schema'> got <class 'dict'>.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\gen-ai\agents_poc\semantic_kernel\venv\Lib\site-packages\proto\marshal\rules\message.py", line 36, in to_proto
    return self._descriptor(**value)
           ~~~~~~~~~~~~~~~~^^^^^^^^^
ValueError: Protocol message Schema has no "anyOf" field.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "google_check3.py", line 69, in <module>
    asyncio.run(main())
ValueError: Unknown field for Schema: anyOf
Platform

Language: Python 3.13
Source: pip package (please run pip show semantic-kernel to get exact version)
AI model: Google AI Gemini 2.0 Flash
IDE: VS Code
OS: Windows

Additional context
This error occurs specifically when using the Google AI connector with agent plugins. The issue appears to be that Semantic Kernel generates JSON schemas containing "anyOf" fields, but Google AI's protobuf Schema definition doesn't support these fields.
The error occurs in the schema conversion process at:

File: google\generativeai\types\content_types.py
Function: FunctionDeclaration.__init__()
During: _rename_schema_fields() processing

This issue prevents using the agent plugin architecture with Google AI services, while other providers (OpenAI, Azure) may work correctly. A schema transformation layer or validation step may be needed to make Semantic Kernel compatible with Google AI's function declaration requirements.

### PR DESCRIPTION
**Pull request by weiguangli-io**
### Motivation and Context

Closes #12442

When `ChatCompletionAgent` instances are used as plugins with `GoogleAIChatCompletion`, the application crashes with:
```
ValueError: Unknown field for Schema: anyOf
```

**Root cause**: The Semantic Kernel JSON schema builder generates `anyOf` for Union types (e.g. `str | list[str]` on the agent's `messages` parameter) and type-as-array for Optional types (e.g. `["string", "null"]`). Google AI's protobuf `Schema` does not support these constructs.

### Description

Add `sanitize_schema_for_google_ai()` to `shared_utils.py` that recursively rewrites unsupported schema constructs:

- `anyOf` / `oneOf` with a null variant → single type + `nullable: true`
- `anyOf` / `oneOf` without null → picks the first variant
- `type` as array (e.g. `["string", "null"]`) → single type + `nullable: true`

Applied in both `kernel_function_metadata_to_google_ai_function_call_format()` and `kernel_function_metadata_to_vertex_ai_function_call_format()`.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The existing tests pass, and I have added 12 new regression tests
- [x] I didn't break any existing functionality

> [!NOTE]
> AI-assisted contribution: Claude helped with code review, test generation, and formatting.

### CODE REVIEW

**Review by moonbox3 — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 3 | **Confidence:** 88%

### ✓ Correctness
> The diff adds a `sanitize_schema_for_google_ai` helper that recursively rewrites JSON schema constructs (`anyOf`, `oneOf`, array-typed `type`) into Google AI-compatible forms. The implementation is correct for the constructs it handles, the tests are thorough, `deepcopy` prevents mutation of the original schema, and edge cases (all-null variants, empty lists) are handled. The only notable gap is that `allOf` is mentioned in the docstring as unsupported by Google AI but is not actually handled, meaning schemas containing `allOf` would pass through unsanitized and likely cause API errors. This is not a blocking issue since `allOf` rarely appears in Pydantic-generated parameter schemas, but it is a correctness gap worth noting.

### ✓ Security Reliability
> The diff adds a recursive schema sanitizer (`sanitize_schema_for_google_ai`) that rewrites JSON-schema constructs unsupported by Google AI (anyOf, oneOf, list-typed `type`) into compatible forms. The implementation uses `deepcopy` to avoid mutating caller data, handles None inputs, and has reasonable fallbacks for edge cases. No security vulnerabilities (injection, secrets, unsafe deserialization) were found. Minor reliability observations exist around missing `allOf` handling (mentioned in the docstring but not implemented) and lack of recursion-depth protection, but neither is a blocking concern given that schemas originate from controlled kernel function metadata.

### ✓ Test Coverage
> The new `sanitize_schema_for_google_ai` utility is well-tested in isolation with good coverage of the main branches (None, passthrough, type-as-list, anyOf/oneOf, recursion, immutability). However, there are no integration-level tests for the two call-sites (`kernel_function_metadata_to_google_ai_function_call_format` and `kernel_function_metadata_to_vertex_ai_function_call_format`) that actually apply the sanitization — these are the functions where the bug fix matters. Additionally, the docstring advertises `allOf` handling but neither the code nor tests address it, and a few edge cases (all-null variants, description already on chosen variant) are uncovered.

### Suggestions
- The docstring for `sanitize_schema_for_google_ai` states that Google AI does not support `allOf`, but the code does not handle it. If a schema with `allOf` is encountered, the key will pass through unsanitized. Consider adding `allOf` to the handling loop alongside `anyOf`/`oneOf`, or removing it from the docstring to avoid confusion.
- In both `google_ai/services/utils.py` and `vertex_ai/services/utils.py`, the guard `if param.schema_data` before calling `sanitize_schema_for_google_ai` is redundant since the function already handles `None` input. This is not a bug, just unnecessary branching.
- The docstring for `sanitize_schema_for_google_ai` mentions `allOf` as an unsupported construct, but the implementation does not handle it. If a schema with `allOf` is passed, it will be forwarded to Google AI unsanitized and may cause an API error. Consider adding handling similar to `anyOf`/`oneOf`, or at minimum logging a warning.
- Consider adding a recursion-depth guard in `_sanitize_node` to protect against pathologically deep or self-referencing schemas (after deepcopy resolves references) that could hit Python's default recursion limit and raise an unhandled `RecursionError`.
- Add unit tests for `kernel_function_metadata_to_google_ai_function_call_format` and `kernel_function_metadata_to_vertex_ai_function_call_format` that pass in `KernelFunctionMetadata` with problematic `schema_data` (e.g. an `anyOf` with null) and assert the output dict/FunctionDeclaration has a sanitized schema. This validates the actual integration, not just the helper.
- The docstring for `sanitize_schema_for_google_ai` mentions `allOf` but the implementation does not handle it. Either add handling + tests for `allOf`, or remove it from the docstring to avoid misleading callers.
- Add a test for the all-null edge case, e.g. `{"type": ["null"]}` or `{"anyOf": [{"type": "null"}]}`, to cover the fallback-to-`"string"` branches.
- Add a test where the chosen `anyOf` variant already carries its own `description`, verifying the outer `description` does not overwrite it (the `if desc and 'description' not in node` guard).
- The Vertex AI `kernel_function_metadata_to_vertex_ai_function_call_format` always constructs a `parameters` dict even when `metadata.parameters` is empty, unlike the Google AI version which guards with `if metadata.parameters`. Consider testing and aligning this behavior.

---
*Automated review by moonbox3's agents*

File: python/semantic_kernel/connectors/ai/google/shared_utils.py [hunk 1]
```diff
@@ -51,6 +53,67 @@ def format_gemini_function_name_to_kernel_function_fully_qualified_name(gemini_f
...
+        if len(type_val) != len(non_null):
+            node["nullable"] = True
+        node["type"] = non_null[0] if non_null else "string"
+
+    # --- handle ``anyOf`` / ``oneOf`` ---
+    for key in ("anyOf", "oneOf"):
+        variants = node.get(key)
+        if not variants:
+            continue
+        non_null = [v for v in variants if v.get("type") != "null"]
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/google/shared_utils.py:87]: `allOf` is documented as unsupported in the docstring (line 62) but is not included in this loop. If a schema containing `allOf` reaches this code, the key will survive unsanitized and likely be rejected by the Google AI API. Consider adding `"allOf"` to the tuple, or removing it from the docstring if handling it is intentionally deferred.

```suggestion
    for key in ("anyOf", "oneOf", "allOf"):
```

File: python/semantic_kernel/connectors/ai/google/shared_utils.py [hunk 2]
```diff
@@ -51,6 +53,67 @@ def format_gemini_function_name_to_kernel_function_fully_qualified_name(gemini_f
...
+        has_null = len(variants) != len(non_null)
+        if non_null:
+            chosen = _sanitize_node(non_null[0])
+        else:
+            chosen = {"type": "string"}
+        # Preserve description from the outer node
+        desc = node.get("description")
+        node.clear()
+        node.update(chosen)
+        if has_null:
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/google/shared_utils.py:94]: The docstring lists `allOf` as unsupported by Google AI, but only `anyOf` and `oneOf` are handled here. An `allOf` key will silently pass through and may cause a downstream API error. Consider adding `"allOf"` to this loop or logging a warning when it is encountered.

```suggestion
    for key in ("anyOf", "oneOf", "allOf"):
```

File: python/semantic_kernel/connectors/ai/google/shared_utils.py [hunk 3]
```diff
@@ -51,6 +53,67 @@ def format_gemini_function_name_to_kernel_function_fully_qualified_name(gemini_f
     return gemini_function_name
 
 
+def sanitize_schema_for_google_ai(schema: dict[str, Any] | None) -> dict[str, Any] | None:
+    """Sanitize a JSON schema dict so it is compatible with Google AI / Vertex AI.
+
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/google/shared_utils.py:58]: Docstring mentions `allOf` but neither the code nor the tests handle it. If `allOf` schemas can appear in practice (e.g. from Pydantic model inheritance), this is a gap. Consider adding handling or removing the mention.

File: python/semantic_kernel/connectors/ai/google/vertex_ai/services/utils.py [hunk 4]
```diff
@@ -135,12 +136,16 @@ def format_tool_message(message: ChatMessageContent) -> list[Part]:
 
 def kernel_function_metadata_to_vertex_ai_function_call_format(metadata: KernelFunctionMetadata) -> FunctionDeclaration:
     """Convert the kernel function metadata to function calling format."""
+    properties = {}
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/google/vertex_ai/services/utils.py]: Unlike the Google AI version, this always builds a `parameters` dict even when `metadata.parameters` is empty. This behavioural difference should be covered by a test (passing metadata with an empty parameter list) and ideally aligned.

File: python/tests/unit/connectors/ai/google/test_shared_utils.py [hunk 5]
```diff
@@ -94,3 +95,119 @@ def test_collapse_function_call_results_in_chat_history() -> None:
...
+        "anyOf": [
+            {"type": "string"},
+            {"type": "array", "items": {"type": "string"}},
+        ],
+        "description": "The user messages for the agent.",
+    }
+    result = sanitize_schema_for_google_ai(schema)
+    assert "anyOf" not in result
+    assert result["type"] == "string"
+    assert result["description"] == "The user messages for the agent."
```
Comment (moonbox3) [inline python/tests/unit/connectors/ai/google/test_shared_utils.py:215]: Good reproducer for issue #12442, but this only tests the sanitizer in isolation. Consider adding a companion integration test that passes this schema as `param.schema_data` inside a `KernelFunctionMetadata` through `kernel_function_metadata_to_google_ai_function_call_format` (and the Vertex AI equivalent) to prove the end-to-end fix.

File: python/semantic_kernel/connectors/ai/google/shared_utils.py (hunk 1 above)
Comment (weiguangli-io) [reply python/semantic_kernel/connectors/ai/google/shared_utils.py:87]: Done -- added `"allOf"` to the loop so it is now `for key in ("anyOf", "oneOf", "allOf"):`. Also added unit tests for `allOf` with and without null variants. Thanks for the catch\!

File: python/semantic_kernel/connectors/ai/google/shared_utils.py (hunk 2 above)
Comment (weiguangli-io) [reply python/semantic_kernel/connectors/ai/google/shared_utils.py:94]: Addressed in the same change as above -- `"allOf"` is now included in the loop tuple.

File: python/semantic_kernel/connectors/ai/google/shared_utils.py (hunk 3 above)
Comment (weiguangli-io) [reply python/semantic_kernel/connectors/ai/google/shared_utils.py:58]: Fixed -- the code now handles `allOf` to match the docstring. Added tests for `allOf` with and without null variants to cover the new handling.

File: python/semantic_kernel/connectors/ai/google/vertex_ai/services/utils.py (hunk 4 above)
Comment (weiguangli-io) [reply python/semantic_kernel/connectors/ai/google/vertex_ai/services/utils.py]: Good catch. I have guarded the Vertex AI version with `if metadata.parameters:` to align with the Google AI version. Note: Vertex AI's `FunctionDeclaration` does not accept `None` for `parameters` (it crashes internally in `_fix_schema_dict_for_gapic_in_place`), so when there are no parameters it still passes an empty `{"type": "object", "properties": {}, "required": []}` dict. Added a test to verify the empty-params case does not crash.

File: python/tests/unit/connectors/ai/google/test_shared_utils.py (hunk 5 above)
Comment (weiguangli-io) [reply python/tests/unit/connectors/ai/google/test_shared_utils.py:215]: Added integration tests for both `kernel_function_metadata_to_google_ai_function_call_format` and `kernel_function_metadata_to_vertex_ai_function_call_format` that pass `KernelFunctionMetadata` with an `anyOf` schema and verify the output is sanitized end-to-end. Also added empty-parameter tests for both functions.

**Comment by weiguangli-io** [comment]
Hi @moonbox3, I've addressed all your review feedback:

1. Added `allOf` to the sanitization loop alongside `anyOf`/`oneOf` — with corresponding unit tests
2. Added integration tests for both `kernel_function_metadata_to_google_ai_function_call_format` and `kernel_function_metadata_to_vertex_ai_function_call_format` with end-to-end schema sanitization
3. Aligned the Vertex AI empty-params behavior with Google AI version (guarded with `if metadata.parameters`)
4. Added tests for the empty-parameter case

Could you please take another look when you get a chance? Thanks for the thorough review!

**Comment by moonbox3** [comment]
@weiguangli-io there are some pre-commit failures, please have a look.

**Comment by weiguangli-io** [comment]
@microsoft-github-policy-service agree

**Comment by weiguangli-io** [comment]
@microsoft-github-policy-service agree company="Individual"
